### PR TITLE
[#3489] improvement(hive-catalog):  Add user authentication e2e test for Hive catalog

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Add e2e tests to test end-to-end user authentication. 

### Why are the changes needed?

Verify that user authentication works

Fix: #3489

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

Test locally.

### CODE REVIEW

File: dev/docker/kerberos-hive/start.sh [hunk 1]
```diff
@@ -42,6 +42,13 @@ kadmin.local -q "ktadd -norandkey -k ${KRB5_KTNAME} hive/${HOSTNAME}@${FQDN}"
 kadmin.local -q "xst -k /hive.keytab -norandkey hive/${HOSTNAME}@${FQDN}"
 kadmin.local -q "xst -k /cli.keytab -norandkey cli@${FQDN}"
 
+# For Gravitino web server
+echo -e "${PASS}\n${PASS}" | kadmin.local -q "addprinc gravitino_client@${FQDN}"
+kadmin.local -q "ktadd -norandkey -k /gravitino_client.keytab gravitino_client@${FQDN}"
+
+echo -e "${PASS}\n${PASS}" | kadmin.local -q "addprinc HTTP/localhost@${FQDN}"
+kadmin.local -q "ktadd -norandkey -k /gravitino_server.keytab HTTP/localhost@${FQDN}"
```
Comment (jerryshao) [inline dev/docker/kerberos-hive/start.sh:50]: I think you'd better move this docker image upgrading thing to a new PR, follow the process to make a tag, update the doc and others, don't mix them into this PR.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java [hunk 2]
```diff
@@ -201,6 +202,7 @@ private void initKerberosIfNecessary(Map<String, String> conf, Configuration had
             new ScheduledThreadPoolExecutor(
                 1, getThreadFactory(String.format("Kerberos-check-%s", info.id())));
 
+        KerberosName.resetDefaultRealm();
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:205]: What is the purpose of this line?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Tag("gravitino-docker-it")
+public class HiveUserAuthenticationIT extends AbstractIT {
+  private static final Logger LOG = LoggerFactory.getLogger(HiveUserAuthenticationIT.class);
+
+  private static final ContainerSuite containerSuite = ContainerSuite.getInstance();
+
+  private static final String SDK_KERBEROS_PRINCIPAL_KEY = "client.kerberos.principal";
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:53]: I fount that you set this to `AbstrctIT`, can you please point out where do we use this config?

File: dev/docker/kerberos-hive/start.sh (hunk 1 above)
Comment (yuqi1129) [reply dev/docker/kerberos-hive/start.sh:50]: OK

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java (hunk 2 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:205]: `KerberosName.resetDefaultRealm();` will reset the Kerberos realm, some tests use realm `EXAMPLE.COM` and some use `HADOOPKRB`, in the same JVM,  there can't be more than one realm here or failures in test will occur.

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 3 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:53]: It was used to transfer the key table and principal for the Gravitino client SDK, If we want to get the version of the Gravitino server with Kerberos enabled, we need pass them to `MiniGravitino`, please see:

image

Comment (jerryshao) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:53]: So these two configurations are faked configurations here only for test?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    GravitinoMetalake[] metalakes = adminClient.listMetalakes();
+    Assertions.assertEquals(0, metalakes.length);
+
+    GravitinoMetalake gravitinoMetalake =
+        adminClient.createMetalake(METALAKE_NAME, null, ImmutableMap.of());
+
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(METASTORE_URIS, HIVE_METASTORE_URI);
+    properties.put(IMPERSONATION_ENABLE, "true");
+    properties.put(KET_TAB_URI, TMP_DIR + HIVE_METASTORE_CLIENT_KEYTAB);
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:179]: Typo: KEY

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    properties.put(PRINCIPAL, HIVE_METASTORE_CLIENT_PRINCIPAL);
+
+    properties.put(CATALOG_BYPASS_PREFIX + HADOOP_SECURITY_AUTHENTICATION, "kerberos");
+    properties.put(
+        CATALOG_BYPASS_PREFIX + "hive.metastore.kerberos.principal",
+        "hive/_HOST@HADOOPKRB"
+            .replace("_HOST", containerSuite.getKerberosHiveContainer().getHostName()));
+    properties.put(CATALOG_BYPASS_PREFIX + "hive.metastore.sasl.enabled", "true");
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:187]: Why do we need to enable sasl here?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            Transforms.EMPTY_TRANSFORM,
+            Distributions.NONE,
+            SortOrders.NONE);
+  }
+
+  private static Column[] createColumns() {
+    Column col1 = Column.of(HIVE_COL_NAME1, Types.ByteType.get(), "col_1_comment");
+    Column col2 = Column.of(HIVE_COL_NAME2, Types.DateType.get(), "col_2_comment");
+    Column col3 = Column.of(HIVE_COL_NAME3, Types.StringType.get(), "col_3_comment");
+    return new Column[] {col1, col2, col3};
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:227]: I think you'd better add more tests, like drop table, create and drop schema.

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 3 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:53]: No, the real value of the Gravitino client SDK is what it uses.

Comment (jerryshao) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:53]: I mean the key is faked...

Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:53]: Yeah

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:187]: The Hive metastore has enabled this configuration. 

https://github.com/datastrato/gravitino/blob/29a3d62ee2a80e38d2a91ed7a71951e185faa4cf/dev/docker/kerberos-hive/hive-site.xml#L59-L61

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 4 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:179]: OK, `KET_TAB_URI` was the value created in the main, let me change it.

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 6 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:239]: added

**Comment by jerryshao** [comment]
Ping me when it is ready for review.

**Comment by yuqi1129** [comment]
> Ping me when it is ready for review.

Got it.

**Comment by yuqi1129** [comment]
> Ping me when it is ready for review.

It's now ready for review.

**Comment by yuqi1129** [comment]
@qqqttt123 
Could you also review it for me?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 5 above)
Comment (jerryshao) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:213]: Why not what you did to enable this?

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    // Now try to alter the table
+    catalog.asTableCatalog().alterTable(tableNameIdentifier, TableChange.rename("new_table"));
+
+    // Drop table
+    catalog
+        .asTableCatalog()
+        .dropTable(NameIdentifier.of(METALAKE_NAME, CATALOG_NAME, SCHEMA_NAME, "new_table"));
+
+    // Drop schema
+    catalog.asSchemas().dropSchema(SCHEMA_NAME, true);
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:257]: Add assertions about what you expected here. not just simply write some codes here, it is a test code, you have to verify something, not blindly throw exceptions if it fails.

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/integration/test/HiveUserAuthenticationIT.java:213]: If we enable Kerberos for Hive, we need to use SSL or SASL security for the transport layer, and compared to SSL, SASL is relatively simple to configure, so we choose to use SASL. That is why the Hive metastore and Hive client need to be enabled.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**